README.md
# 概要

OAuth 認証に必要な Authorization シグネチャを作成するライブラリ。

# 必要条件

| OS      | バージョン |
| :------ | :--------- |
| Ubuntu  | `16.04`    |

| アプリケーション | バージョン               |
| :--------------- | :----------------------- |
| node.js          | `>=8.11.4`               |
| npm              | `>=5.6.0`                |

## OAuthバージョン

| version       |
| :-----------: |
| `1.0`         |

## 対応API

* Twitter API

# スタートガイド

`package.json` の `dependencies` に以下２つの項目をを追記する。
```
"@sevenspice/warbler": "https://github.com/sevenspice/Warbler.git",
"request": "^2.88.0"
```

パッケージをインストールする。
```
npm install
```

ツイートを検索する場合のソースコード例。
```
touch index.js
vi index.js
```

ファイルに下記をコピーアンドペースト。
* コンシューマーキー、コンシューマーシークレット、アクセストークン、アクセストークンシークレットは適宜変更すること。
``` javascript
const Warbler = require('@sevenspice/warbler');
const Request = require('request');

const scheme              = 'https';
const host                = 'api.twitter.com';
const entry_point         = '/1.1/search/tweets.json';
const protocol            = 'GET';
const consumer_key        = 'コンシューマーキー';
const consumer_secret     = 'コンシューマーシークレット';
const access_token        = 'アクセストークン';
const access_token_secret = 'アクセストークンシークレット';
const options   = { q:'#天気', locale:'ja' };
const timeout   = 10000;
const timestamp = Math.round((new Date()).getTime() / 1000);

const walber = new Warbler(
    scheme
    , host
    , entry_point
    , protocol
    , consumer_key
    , consumer_secret
    , access_token
    , access_token_secret
    , options, timeout
);

const getResponse = (_, __, body) => {
    const tweets   = JSON.parse(body);
    const statuses = tweets.statuses;
    for(let i = 0; i < statuses.length; i++){
        process.stdout.write(`${statuses[i].id}\n`);
        process.stdout.write(`${statuses[i].text}\n`);
    }
};

Request.get({
    url: `${walber.getScheme()}://${walber.getHost()}${walber.getEntryPoint()}`,
    qs: walber.getOptions(),
    headers: {
        'Authorization':  walber.getAuthString(timestamp, walber.getNonce())
        , 'content-type': 'application/x-www-form-urlencoded'
    },
}, getResponse);

```

スクリプトを実行。
```
node index.js
```

# Direct Messageについて

Direct Message では若干他のAPIと仕様が若干異なるため注意が必要。以下に送信例を示す。
* [Sending and receiving events](https://developer.twitter.com/en/docs/direct-messages/sending-and-receiving/overview)

送信コード例
``` javascript
const Warbler = require('@sevenspice/warbler');
const Request = require('request');

const scheme              = 'https';
const host                = 'api.twitter.com';
const entry_point         = '/1.1/direct_messages/events/new.json';
const protocol            = 'POST';
const consumer_key        = 'コンシューマーキー';
const consumer_secret     = 'コンシューマーシークレット';
const access_token        = 'アクセストークン';
const access_token_secret = 'アクセストークンシークレット';
// options には DM に対応した JSON オブジェクトを指定する
const options   = { event: { type: "message_create", message_create: { target: { recipient_id: "USER_ID" } , message_data: { text: 'すごーい！' } } } };
const timeout   = 10000;
const timestamp = Math.round((new Date()).getTime() / 1000);

const walber = new Warbler(
    scheme
    , host
    , entry_point
    , protocol
    , consumer_key
    , consumer_secret
    , access_token
    , access_token_secret
    , options, timeout
);

const getResponse = (_, __, body) => {
    console.log(body);
};

request = Request.post({
    url: `${walber.getScheme()}://${walber.getHost()}${walber.getEntryPoint()}`,
    json: walber.getOptions(), // HTTP 通信に request パッケージを使用している場合は、qs ではなく json を指定する
    headers: {
        'Authorization':  walber.getAuthString(timestamp, walber.getNonce())
        , 'content-type': 'application/json' // application/x-www-form-urlencoded ではなく application/json を指定する
    },
}, getResponse);

```
* `recipient_id`は送りたいユーザーのIDを指定する。
* ユーザーのIDはエントリポイントの`/1.1/users/show.json`を使用して調べることができる。
    * [GET users/show](https://developer.twitter.com/en/docs/accounts-and-users/follow-search-get-users/api-reference/get-users-show)

取得コード例
``` javascript
const Warbler = require('@sevenspice/warbler');
const Request = require('request');

const scheme              = 'https';
const host                = 'api.twitter.com';
const entry_point         = '/1.1/direct_messages/events/list.json';
const protocol            = 'GET';
const consumer_key        = 'コンシューマーキー';
const consumer_secret     = 'コンシューマーシークレット';
const access_token        = 'アクセストークン';
const access_token_secret = 'アクセストークンシークレット';
const options   = {};
const timeout   = 10000;
const timestamp = Math.round((new Date()).getTime() / 1000);

const walber = new Warbler(
    scheme
    , host
    , entry_point
    , protocol
    , consumer_key
    , consumer_secret
    , access_token
    , access_token_secret
    , options, timeout
);

const getResponse = (_, __, body) => {
    console.log(body);
};

request = Request.get({
    url: `${walber.getScheme()}://${walber.getHost()}${walber.getEntryPoint()}`,
    headers: {
        'Authorization':  walber.getAuthString(timestamp, walber.getNonce())
        , 'content-type': 'application/json'
    },
}, getResponse);

```

CHANGELOG.md
# 1.1.0 - 2020/3/20
## 追加
* Direct Message API 対応。
* README.mdの更新。


sources/spec/test_warbler.js
'use strict';
/* global describe   */
/* global before     */
/* global beforeEach */
/* global it         */
/* global afterEach  */
/* global after      */

const chai    = require('chai');
const assert  = chai.assert;
const Warbler = require(`${__dirname}/../main/warbler`);

describe('Warbler class test', () => {

    const scheme              = 'https';
    const host                = 'api.twitter.com';
    const entry_point         = '/1.1/search/tweets.json';
    const protocol            = 'GET';
    const consumer_key        = 'v0s9tO9WrMHh3yEnhqPujKSag1v6VMIc';
    const consumer_secret     = 'jr7HtnJF1VJN7iG65ibMNh13H2kaLqZ9';
    const access_token        = '8aGkj7x720sxrixbCi0KFilOysGmXSgI';
    const access_token_secret = 'mSBi1SW73czsajiCxVhHQ4jPHx9fEIqQ';
    let warbler = null;

    before( () => {});

    /*
     * 各テスト実施前に実行される
     */
    beforeEach( () => {});

    it('instance', () => {

        const options = {q:'twitter', locale:'ja'};

        // 引数をすべて指定してインスタンス化
        warbler = new Warbler(
            scheme, host, entry_point, protocol, consumer_key
            ,consumer_secret
            ,access_token
            ,access_token_secret
            ,options, 10000
        );


        assert.equal('object', (typeof warbler), 'warbler is not object.');
        assert.equal(scheme,       warbler.getScheme());
        assert.equal(host,         warbler.getHost());
        assert.equal(entry_point,  warbler.getEntryPoint());
        assert.equal(protocol,     warbler.getProtocol());
        assert.equal(consumer_key, warbler.getConsumerKey());
        assert.equal(access_token, warbler.getAccessToken());
        assert.equal(`${consumer_secret}&${access_token_secret}`, warbler.getSecretKey());
        assert.equal(options.q,      warbler.getOptions().q);
        assert.equal(options.locale, warbler.getOptions().locale);
        assert.equal(10000, warbler.getTimeout());


        // クエリオプションとタイムアウトを指定せずにインスタンス化
        warbler = null;
        warbler = new Warbler(
            scheme, host, entry_point, protocol, consumer_key
            ,consumer_secret
            ,access_token
            ,access_token_secret
        );


        assert.equal('object', (typeof warbler), 'warbler is not object.');
        assert.equal(scheme,       warbler.getScheme());
        assert.equal(host,         warbler.getHost());
        assert.equal(entry_point,  warbler.getEntryPoint());
        assert.equal(protocol,     warbler.getProtocol());
        assert.equal(consumer_key, warbler.getConsumerKey());
        assert.equal(access_token, warbler.getAccessToken());
        assert.equal(`${consumer_secret}&${access_token_secret}`, warbler.getSecretKey());
        assert.equal(0,    Object.keys(warbler.getOptions()).length);
        assert.equal(5000, warbler.getTimeout());

        warbler.setOptions(options);
        assert.notEqual(0, Object.keys(warbler.getOptions()).length);
        assert.equal(options.q,      warbler.getOptions().q);
        assert.equal(options.locale, warbler.getOptions().locale);

        warbler.setTimeout(20000);
        assert.equal(20000, warbler.getTimeout());


    });

    it('getOptionsUrlEncoded', () => {

        warbler.setOptions({status:'艦これ'});
        const options = warbler.getOptionsUrlEncoded();

        assert.equal('%E8%89%A6%E3%81%93%E3%82%8C', options['status']);

    });

    it('getAuthString', () => {

        // 返却されるはずの文字列
        const comparison = 'OAuth oauth_consumer_key="6TrtNCAdrFounnGuensSQ1a6a", oauth_nonce="372cb60308b5f0479cda91149db21022", oauth_signature="zLDk31PcF6bXPVhy1LaR%2FLkD8zs%3D", oauth_signature_method="HMAC-SHA1", oauth_timestamp="1505374547", oauth_token="822787074-eNcVAnBjdn1TTZ4t1nY6vMZBf8ISKB0FhKWBeOzX", oauth_version="1.0", status="%25E8%2589%25A6%25E3%2581%2593%25E3%2582%258C"';

        const protocol            = 'POST';
        const entry_point         = '/1.1/statuses/update.json';
        const consumer_key        = '6TrtNCAdrFounnGuensSQ1a6a';
        const consumer_secret     = 'TWWdXzvbsQICL2NotMcIcPxcLDsl7siIwbpaggVU2KcxWk6mgI';
        const access_token        = '822787074-eNcVAnBjdn1TTZ4t1nY6vMZBf8ISKB0FhKWBeOzX';
        const access_token_secret = 'bNHFjslVpXFN6MB4kW7OuhRQ2ZJ1IaBSztXmgm4Z9a0PC';
        const timestamp           = '1505374547';
        const nonce               = '372cb60308b5f0479cda91149db21022';
        const options             = {status:'艦これ'};

        const w = new Warbler(
            scheme, host, entry_point, protocol, consumer_key
            ,consumer_secret, access_token, access_token_secret
            ,options, 10000
        );

        const result = w.getAuthString(timestamp, nonce);

        assert.equal(comparison, result);

    });

    it('sign', () => {

        const target = 'Hello World';
        const key    = 'secret';

        const w = new Warbler(
            scheme, host, entry_point, protocol, consumer_key
            ,key, access_token, '' ); 

        const result = w.sign(target);
        assert.equal('Rn0mSLYlmbiWaLspobTbLILIkZs=', result);

    });

    it('getSignatureString', () => {

        // 返却されるはずの文字列
        const comparison = 'POST&https%3A%2F%2Fapi.twitter.com%2F1.1%2Fstatuses%2Fupdate.json&oauth_consumer_key%3Dxvz1evFS4wEEPTGEFPHBog%26oauth_nonce%3DkYjzVBB8Y0ZFabxSWbWovY3uYSQ2pTgmZeNu2VS4cg%26oauth_signature_method%3DHMAC-SHA1%26oauth_timestamp%3D1318622958%26oauth_token%3D370773112-GmHxMAgYyLbNEtIKZeRNFsMKPR9EyMZeS9weJAEb%26oauth_version%3D1.0%26status%3D%25E8%2589%25A6%25E3%2581%2593%25E3%2582%258C';

        const protocol     = 'POST';
        const entry_point  = '/1.1/statuses/update.json';
        const consumer_key = 'xvz1evFS4wEEPTGEFPHBog';
        const access_token = '370773112-GmHxMAgYyLbNEtIKZeRNFsMKPR9EyMZeS9weJAEb';
        const nonce        = 'kYjzVBB8Y0ZFabxSWbWovY3uYSQ2pTgmZeNu2VS4cg';
        const timestamp    = '1318622958';
        const options      = {status:'艦これ'};

        const w = new Warbler(

            scheme, host, entry_point, protocol, consumer_key
            ,consumer_secret, access_token, access_token_secret
            ,options, 10000

        );

        const result = w.getSignatureString(timestamp, nonce);

        assert.equal(comparison, result);

    });


    it('getNonce', () => {

        const nonce = warbler.getNonce();

        // 32桁であること
        assert.equal(32, nonce.length);

        const nonce1 = warbler.getNonce();
        const nonce2 = warbler.getNonce();

        // 同じ文字列が作成されないこと
        assert.notEqual(nonce1, nonce2);


    });

    it('getQueryMap', () => {

        const nonce = warbler.getNonce();
        const timestamp = Math.round((new Date()).getTime() / 1000);
        const signature = 'abcdefghjklmnopq';
        let keys = [];

        warbler.setOptions({q:'twitter', locale:'ja'});
        let map = warbler.getQueryMap(timestamp, nonce, signature);

        // 要素数は9
        assert.equal(9, Object.keys(map).length);

        for(const key in map){
            keys.push(key);
        }

        // キーの並び順が昇順であること
        assert.equal('locale',                 keys[0]);
        assert.equal('oauth_consumer_key',     keys[1]);
        assert.equal('oauth_nonce',            keys[2]);
        assert.equal('oauth_signature',        keys[3]);
        assert.equal('oauth_signature_method', keys[4]);
        assert.equal('oauth_timestamp',        keys[5]);
        assert.equal('oauth_token',            keys[6]);
        assert.equal('oauth_version',          keys[7]);
        assert.equal('q',                      keys[8]);

        // キーと値の対応が正しいこと
        assert.equal('ja',         map[keys[0]]);
        assert.equal(consumer_key, map[keys[1]]);
        assert.equal(nonce,        map[keys[2]]);
        assert.equal(signature,    map[keys[3]]);
        assert.equal('HMAC-SHA1',  map[keys[4]]);
        assert.equal(timestamp,    map[keys[5]]);
        assert.equal(access_token, map[keys[6]]);
        assert.equal('1.0',        map[keys[7]]);
        assert.equal('twitter',    map[keys[8]]);

        keys = [];
        // 署名は指定しない
        map = warbler.getQueryMap(timestamp, nonce);

        // 要素数は8
        assert.equal(8, Object.keys(map).length);

        for(const key in map){
            keys.push(key);
        }

        // キーの並び順が昇順であること
        assert.equal('locale',                 keys[0]);
        assert.equal('oauth_consumer_key',     keys[1]);
        assert.equal('oauth_nonce',            keys[2]);
        assert.equal('oauth_signature_method', keys[3]);
        assert.equal('oauth_timestamp',        keys[4]);
        assert.equal('oauth_token',            keys[5]);
        assert.equal('oauth_version',          keys[6]);
        assert.equal('q',                      keys[7]);

        // キーと値の対応が正しいこと
        assert.equal('ja',         map[keys[0]]);
        assert.equal(consumer_key, map[keys[1]]);
        assert.equal(nonce,        map[keys[2]]);
        assert.equal('HMAC-SHA1',  map[keys[3]]);
        assert.equal(timestamp,    map[keys[4]]);
        assert.equal(access_token, map[keys[5]]);
        assert.equal('1.0',        map[keys[6]]);
        assert.equal('twitter',    map[keys[7]]);

        // DM対応
        // DMのエントリポイントを指定されている場合はJSONはクエリーパラメーターに含めない
        const w = new Warbler(
            scheme
            ,host
            ,'/1.1/direct_messages/events/new.json'
            ,'POST'
            ,consumer_key
            ,consumer_secret
            ,access_token
            ,access_token_secret
            ,{event: {type: "message_create", message_create: {target: {recipient_id: "822787074"} , message_data: {text: 'Hello,World'}}}}
            ,10000
        );

        keys = [];
        map = w.getQueryMap(timestamp, nonce);

        for(const key in map){
            keys.push(key);
        }

        // JSONのキーが含まれていないこと
        assert.isFalse(keys.includes('type'));
        assert.isFalse(keys.includes('message_create'));

    });

    it('sortMapByAsc', () => {

        const sample = { ringo:'apple', budou: 'grape', kaki: 'persimmon', akebi: 'chocolate vine'};
        const sorted = warbler.sortMapByAsc(sample);
        const keys   = [];

        for(const key in sorted){
            keys.push(key);
        }

        // キーの並びが昇順であること
        assert.equal('akebi', keys[0]);
        assert.equal('budou', keys[1]);
        assert.equal('kaki' , keys[2]);
        assert.equal('ringo', keys[3]);

        // キーと値の対応が正しいこと
        assert.equal('chocolate vine', sorted[keys[0]]);
        assert.equal('grape',     sorted[keys[1]]);
        assert.equal('persimmon', sorted[keys[2]]);
        assert.equal('apple',     sorted[keys[3]]);

    });

    /*
     * 各テスト実施後に実行される
     */
    afterEach( () => {});

    after( () => {});

});

sources/main/warbler.js
'use strict';

/**
 * Twitter APIに対応したOAuth認証用の設定値を作成するクラス
 */
class Warbler {
    /**
     * コンストラクタ
     * @param {string}  scheme              スキーム
     * @param {string}  host                Twitter APIのホスト名
     * @param {string}  path                エントリポイント
     * @param {string}  protocol            'GET'または'POST'
     * @param {string}  consumer_key        コンシューマーキー
     * @param {string}  consumer_secret     コンシューマーシークレットキー
     * @param {string}  access_token        アクセストークン
     * @param {string}  access_token_secret アクセストークンシークレット
     * @param {object}  options             クエリに追加するオプション
     * @param {integer} timeout             タイムアウト時間（ミリ秒）
     */
    constructor(

        scheme, host, entry_point, protocol, consumer_key, consumer_secret
        ,access_token, access_token_secret, options = {}, timeout = 5000

    ){

        this.scheme = scheme;
        this.host   = host;
        this.entry_point  = entry_point;
        this.protocol     = protocol;
        this.consumer_key = consumer_key;
        this.access_token = access_token;
        this.secret_key   = `${consumer_secret}&${access_token_secret}`;
        this.options = options;
        this.timeout = timeout;

    }

    /**
     * スキームを取得する
     * @return {string} スキームを返却
     */
    getScheme () {
        return this.scheme;
    }

    /**
     * ホストを取得する
     * @return {string} ホストを返却
     */
    getHost () {
        return this.host;  
    }

    /**
     * エントリポイントを取得する
     * @reutrn {string} エントリポイントを返却
     */
    getEntryPoint () {
        return this.entry_point;
    }

    /**
     *  プロトコルを取得する
     * @return {string} プロトコルを返却
     */
    getProtocol () {
        return this.protocol;
    }

    /**
     * コンシューマーキーを取得する
     * @return {string} コンシューマーキーを返却
     */
    getConsumerKey () {
        return this.consumer_key;
    }

    /**
     * アクセストークンを取得する
     * @return {string} アクセストークンを返却
     */
    getAccessToken () {
        return this.access_token;
    }

    /**
     * シークレットキーを取得する
     * @return {string} シークレットキーを返却
     */
    getSecretKey () {
        return this.secret_key;
    }

    /**
     * クエリに追加するオプションを取得する
     * @return {object} クエリに追加するオプションを返却
     */
    getOptions () {
        return this.options;
    }

    /**
     * URLエンコードされたクエリに追加するオプションを取得する
     * @return {object} クエリに追加するオプションを返却（URLエンコード済み）
     */
    getOptionsUrlEncoded () {

        const queries = {};

        for(const key in this.getOptions()){
            queries[key] = encodeURIComponent(this.getOptions()[key]);
        }

        return queries;

    }

    /**
     * クエリに追加するオプションを設定する
     * @param {object} options クエリに追加するオプションを返却
     */
    setOptions (options) {
        this.options = options;
    }

    /**
     * タイムアウト時間を取得する
     * @return {integer} タイムアウト時間（ミリ秒）を返却
     */
    getTimeout () {
        return this.timeout;
    }

    /**
     * タイムアウト時間を設定する
     * @param {integer} timeout タイムアウト時間（ミリ秒）を返却
     */
    setTimeout (timeout) {
        this.timeout = timeout;
    }

    /**
     * クラス定数
     */
    static get CODE_TABLE () {
        return '0123456789abcdef';
    }

    /**
     * クラス定数
     */
    static get DIRECT_MESSAGE_REGULAR_EXPRESSION () {
        return '/direct_messages/';
    }

    /**
     * 認証に使用する文字列を生成して返却する
     * @param {string} timestamp 送信時のタイムスタンプ値
     * @param {string} nonce     ランダム文字列
     * @return {string} HTTPヘッダーのAuthorizationキーに対応する認証用文字列を返却
     */
    getAuthString (timestamp, nonce) {

        let auth = 'OAuth ';

        // 署名を取得
        const signature = this.sign(this.getSignatureString(timestamp, nonce));

        // 署名をクエリーオブジェクトに追加
        const map = this.getQueryMap(timestamp, nonce, signature);

        for(const key in map){
            auth += `${key}="${encodeURIComponent(map[key])}", `;
        }

        // 末尾の,を削除して返却
        return auth.slice(0, -2);

    }

    /**
     * 署名を生成して返却する
     * ＊ハッシュはsha1で生成
     * @param {string} 署名元の文字列を指定
     * @return {string} 署名を返却（base64でエンコード済み）
     */
    sign (target) {

        const crypto = require('crypto');
        return crypto.createHmac('sha1', this.getSecretKey()).update(target).digest('base64');

    }

    /**
     * 署名のための文字列を生成して返却する
     * @param {string} timestamp 送信時のタイムスタンプ値
     * @param {string} nonce     ランダム文字列
     * @param {string} signature OAuth認証用の署名
     * @return {string} 署名用文字列を返却
     */
    getSignatureString (timestamp, nonce) {

        const url = `${this.getScheme()}://${this.getHost()}${this.getEntryPoint()}`;
        const map = this.getQueryMap(timestamp, nonce);
        let base  = `${this.getProtocol()}&${encodeURIComponent(url)}&`;

        for(const key in map){
            base += encodeURIComponent((key + '=' + map[key] + '&'));
        }

        // 末尾の&を削除して返却
        return base.slice(0, -3);

    }

    /**
     * 送信するクエリーパラメーターを内包するオブジェクトを返却する
     * @param {string} timestamp 送信時のタイムスタンプ値
     * @param {string} nonce     ランダム文字列
     * @param {string} signature OAuth認証用の署名
     * @return {object} クエリパラメーターを内包するオブジェクトを返却
     */
    getQueryMap (timestamp, nonce, signature = '') {

        const parameters = {

            'oauth_token':        this.getAccessToken(),
            'oauth_consumer_key': this.getConsumerKey(),
            'oauth_timestamp': timestamp, 'oauth_signature_method': 'HMAC-SHA1',
            'oauth_nonce': nonce, 'oauth_version': '1.0'

        };

        // DM対応
        // DM時にはクエリをパラメーターに加えない
        if( ! this.getEntryPoint().match(Warbler.DIRECT_MESSAGE_REGULAR_EXPRESSION)){
            for(const key in this.getOptions()){
                parameters[key] = encodeURIComponent(this.getOptions()[key]);
            }
        }

        // 署名追加
        if(signature != '') parameters['oauth_signature'] = signature;

        // キーを基準に昇順に並べ替えて返却
        return this.sortMapByAsc(parameters);

    }

    /**
     * NONCE（使い捨てのランダムな値のこと）を生成して返却する
     * @return {string} ３２桁の文字列を返却
     */
    getNonce () {

        let   nonce = '';
        const digit = (Warbler.CODE_TABLE.length * 2);

        for(let i=0; i<digit; i++){
            nonce += Warbler.CODE_TABLE.charAt(Math.floor((Math.random() * Warbler.CODE_TABLE.length)));
        }

        return nonce;

    }

    /**
     * オブジェクトの中身をキーに対して昇順で並べ替える
     * @param {object} 中身を並べ替えるオブジェクト
     * @return {object} 並べ替えたオブジェクトを返却
     */
    sortMapByAsc (object) {

        const sorted = {};
        const keys   = [];
 
        // 引数に渡されたオブジェクトのキーのみを取り出す
        for(const key in object){
            keys.push(key);
        }

        // 昇順にソート
        keys.sort();

        // 再格納
        for(let i=0; i< keys.length; i++){
            sorted[keys[i]] = object[keys[i]];
        }

        return sorted;

    }

}

// エクスポート
module.exports = Warbler;


api.sh
#!/bin/bash

# -------------------------------------------- #
# APIへのアクセスをテストするシェルスクリプト
# -------------------------------------------- #

# POST 投稿（-dは「艦これ」をURLエンコードした文字列）
curl 'https://api.twitter.com/1.1/statuses/update.json' \
-H 'application/x-www-form-urlencoded' \
-H 'Authorization: OAuth oauth_consumer_key="9TrtNCAdrFounnGxegsSQ1a6n", oauth_nonce="372cb60308b5f0479cda91149db21022", oauth_signature="lpfZJmTmdxCS3aDxUsuRLkumchw%3D", oauth_signature_method="HMAC-SHA1", oauth_timestamp="1505374547", oauth_token="822787099-eNcVAnBjdn1TTZ4t1nY6vMZDg9HSKB0FhKWBeOyU", oauth_version="1.0", status="%25E8%2589%25A6%25E3%2581%2593%25E3%2582%258C"' \
-X POST \
-d 'status=%E8%89%A6%E3%81%93%E3%82%8C' \
-v

# GET 検索
#curl \
#--verbose \
#--header 'application/x-www-form-urlencoded' \
#--header 'Authorization: OAuth oauth_consumer_key="6TrtNCAdrFozummGuensSQ1a9n", oauth_nonce="df8f204cbc2f4ad257ec3ef397e90f3a", oauth_signature="QtETcBu3BZEcpg51OyAcohIVaOo%3D", oauth_signature_method="HMAC-SHA1", oauth_timestamp="1505372700", oauth_token="833787074-eNcVAnBjdn1TTZ4t1bY6vXXBg8ISKB0FhKWBeOzP", oauth_version="1.0", q="%2523twitter"' \
#'https://api.twitter.com/1.1/search/tweets.json?q=%23twitter'
